### ISSUE TITLE
Cypress cleanup

### ISSUE THREAD

**Issue by nrfox** [kiali/kiali#5069]
This is a catch-all issue for better organizing the cypress tests and feature files based on learning from #4854. It should be worked on only after #4854 is finished as getting coverage in place is a higher priority but this issue is to track potential improvements we've discovered along the way.

- [ ] Tests should be written in a more Gherkin friendly syntax and less of a procedural style
- [ ] Document best practices for writing tests
- [ ] Test cross cutting features with tags

**Comment by lucasponce** [comment]
Before planning any refactoring of the existing tests, please wait until QE has a plan for a massive migration from legacy Selenium to cypress.

Some conversations are ok and good, but I think we are missing some context here.

Please could re-write the same tests one and another just for style and procedures and that's not the goal of this effort.

**Comment by nrfox** [comment]
> Please could re-write the same tests one and another just for style and procedures and that's not the goal of this effort.

Agreed that the immediate goal should be to get more coverage and before any time is spent re-writing existing tests that are working, we should work on getting more coverage. Also the syntax of the tests shouldn't be a blocker. Ideally, all the tests would be written in a similar style that follows Gherkin recommendations but that shouldn't be the priority right now. This issue is just to capture some of the ways we'd like to improve the test suite in the future based on learnings from writings tests during #4854.

**Comment by israel-hdez** [comment]
I heard that Selenium was phased out.
@nrfox I wonder if this still holds and can be itemized, or if we can close this?

**Comment by nrfox** [comment]
> I heard that Selenium was phased out. @nrfox I wonder if this still holds and can be itemized, or if we can close this?

This issue isn't related to Selenium. It's about cypress improvements we'd like to make down the road but we didn't prioritize during the migration. imo the most important issue for cypress refactoring is addressing step definition re-usability. The way that most of the tests are currently organized, with a separate `.ts` file for each test, does not make it easy when writing a new test to discover common definitions that you could've re-used. Instead we end up re-implementing much of the same functionality in new definitions and we end up with many different variations of the same step definitions. This makes flakes more likely since you can address a cause for a flake in one step definition without realizing there are 2-3 other similar definitions out there. There's also no namespacing of the step definitions so we'll either end up with many variations of the same definitions or we'll need to refactor/organize them in a better way. I'd like to see them organized roughly by function, i.e. all the graph related definitions in one file, all the navigation related definitions in another. There's bound to be some overlap and we'll probably want to further subdivide things but it's at least a place to start. Open to suggestions and would like to hear what others think as well on how we can organize things better. For reference [here's what the gherkin docs suggest](https://cucumber.io/docs/gherkin/step-organization/?lang=javascript) for organizing these.

**Comment by israel-hdez** [comment]
> This issue isn't related to Selenium. It's about cypress improvements

Yeah, I remember. I mentioned Selenium because I think we can expect (and indeed I'm seeing) more activity on adding cypress tests to increase coverage. If the test suite keeps growing with the same organization, the point where the effort won't justify the value will be reached.

The `future` label is still there. However, I'm starting to see this as "do now, or never do it". This is why I'm wondering if we should itemize/do this, or simply close it.

**Comment by jshaughn** [comment]
We have a new-hire pending, this could be a good initial, although not easy, task to help become familiar with the project.

**Comment by israel-hdez** [comment]
@jshaughn Well, that is true. It can wait a little bit :-)

**Comment by jshaughn** [comment]
We have added tags and improved organization.  I think this catch-all can be replaced with explicit improvement tasks going forward.
